### ISSUE TITLE
Fail to check signature

### ISSUE THREAD

**Issue by lucasdemarchi** [rpm-software-management/rpm#270]
This happens when updating update from 4.13.0.1-release to master branch:

```sh
rpm -qlp /tmp/capsule-01.00.12-r0.core2_64.rpm
error: /tmp/capsule-01.00.12-r0.core2_64.rpm: signature region 62: tag number mismatch 7 ril 6 dl 792 rdl 168

error: /tmp/capsule-01.00.12-r0.core2_64.rpm: not an rpm package (or package manifest)
```

I used commit a8e51b3bb05c6acb1d9b2e3d34f859ddda1677be as  starting point for a bisect since I knew it was failing and was already with it checked out (related bug: https://github.com/intel-aero/meta-intel-aero/issues/219).

Test script:
```sh
cat /tmp/test.sh 
#!/bin/bash -e

git clean -ffxd
./autogen.sh  --prefix=/usr --without-lua --disable-python --with-external-db \
    CPPFLAGS="-I/usr/include/nspr4 -I/usr/include/nss3 -I/usr/include/db45 "
make -j4 
./rpm -qlp /tmp/capsule-01.00.12-r0.core2_64.rpm
```

Bisect log:
```
# Tested with  with "git bisect run /tmp/test.sh"


git bisect start
# bad: [a8e51b3bb05c6acb1d9b2e3d34f859ddda1677be] brp-python-bytecompile: Process python lib dirs even without standard Python
git bisect bad a8e51b3bb05c6acb1d9b2e3d34f859ddda1677be
# good: [e97d475ae2cf6709f2bbdaba34f90b87d454c13c] Preparing for rpm 4.13.0.1
git bisect good e97d475ae2cf6709f2bbdaba34f90b87d454c13c
# good: [d27f0167aa53951c4164a5d49149ad416c5a1ee4] Get rid of warnigs caused by d18fdd7f8964d01b4dfc39bf84133695cff5a13a
git bisect good d27f0167aa53951c4164a5d49149ad416c5a1ee4
# good: [e5d3b9f682383ee026b5a86a47eb5f288bdf8991] rpmdb.c: (rpmdbCheckTerminate) return non-zero on subsequent runs
git bisect good e5d3b9f682383ee026b5a86a47eb5f288bdf8991
# bad: [5b714956b7776edea635f73daa1137aa44609290] Optionally validate string is hex in rpmSigInfoParse()
git bisect bad 5b714956b7776edea635f73daa1137aa44609290
# bad: [e9e637e2559926ef0d4aa3f7cf686d1944e1d036] Fix expandFormat() to work with all string types
git bisect bad e9e637e2559926ef0d4aa3f7cf686d1944e1d036
# good: [cb97c0161236e5ba9d560b95cadd41eba87cccfc] Add + use a helper function for converting entry info to host endianess
git bisect good cb97c0161236e5ba9d560b95cadd41eba87cccfc
# bad: [0cc5ee83f1f7f345a9d46166eac2cbe4d5ba84dc] Bury the last remains of non-header signatures into lead
git bisect bad 0cc5ee83f1f7f345a9d46166eac2cbe4d5ba84dc
# good: [d26dff0632f5a23c6da3795656e8137d4057bc13] Eliminate unnecessary entry argument to headerVerifyRegion()
git bisect good d26dff0632f5a23c6da3795656e8137d4057bc13
# good: [66145664fb1d7bb7e9d584c69f97c7869de5997f] Whitespace cleanup in headerVerifyRegion()
git bisect good 66145664fb1d7bb7e9d584c69f97c7869de5997f
# bad: [441a0998fca9b3d01052bc3d33f8a72f725d5865] Require exact match on header vs region sizes when reading package files
git bisect bad 441a0998fca9b3d01052bc3d33f8a72f725d5865
# good: [6e0b5dc9f10d7d3ee944dcd63d1c6019b2e80f0d] Verify region data size fits the header
git bisect good 6e0b5dc9f10d7d3ee944dcd63d1c6019b2e80f0d
# first bad commit: [441a0998fca9b3d01052bc3d33f8a72f725d5865] Require exact match on header vs region sizes when reading package files
```

Test rpm can be downloaded from here: https://downloadmirror.intel.com/26500/eng/capsule-01.00.12-r0.core2_64.rpm

It's not obvious for me what the fix should be since the commit itself should actually be a fix as far as I understand. Doing the wrong thing and removing the check for `exact_size` "fixes" the issue, but is also probably the wrong thing to do.

**Comment by n3npq** [comment]
That is a package produced by RPM5:
```
$ /opt/local/bin/rpm -qp --yaml ./capsule-01.00.12-r0.core2_64.rpm | grep Rpmversion
  Rpmversion: 5.4.16
```

The package uses a non-repudiable signature and verifies fine:

```
$ /opt/local/bin/rpm -qvvp  ./capsule-01.00.12-r0.core2_64.rpm
...
D:   PUB: 3375C959 31F7927C V4 DSA
D: ========== DSA pubkey id 3375c959 31f7927c (package)
D: ./capsule-01.00.12-r0.core2_64.rpm: Header V4 DSA signature: OK, key ID 31f7927c
```

However, RPM4 can't find the pubkey contained in the package.

**Comment by n3npq** [comment]
Thanks for identifying the specific checkin.

RPM4 has changed the definition of what is called an "immutable region" in a signature header.

An "immutable region" is identified with an index/data length as well as a trailer which duplicates
the same information. The test for an exact_match in the commit has altered the definition
of an "immutable region" and appears to have introduced an incompatibility.

If you are going to handle Yocto packages built by RPM5 using RPM4 sanely, your best bet is to back out that specific "exact_match" change. You will be no worse off than you were before, the test is rather arcane and anal retentive.

**Comment by lucasdemarchi** [comment]
@n3npq thanks for looking into this.  This rpm may be the only one I'll have to deal with and I can regenerate it using rpm4 reasonably easy. However I thought it was important to notify upstream since when other distros update they may face similar problems since they don't have much control over what .rpm users are going to install.

Maybe an option to rpm to allow it to continue even if the "exact_match" test fails? Or changing this to be a warning rather than an error since it causes compatibility issues?

Thanks

**Comment by lucasdemarchi** [comment]
Regarding using RPM4 to manager RPM5 packages, the reason is that the previous version of the distro was using RPM5. However due to incompatibilities with dnf it was decided by uptream Yocto developers do switch to RPM4. In our Yocto-based distro there are some leftover rpm files that weren't rebuilt since it's a BIOS package and there weren't any updates to it.

**Comment by n3npq** [comment]
The often stated public policy here with RPM4 is "Compatibility with RPM5 is a negative goal."
Don't hold your breath for a patch: I will generate and distribute a patch as soon as I understand
the full consequences of recent changes to the RPM format. 

Yes, Yocto issues are well known. The rpm5+dnf incompatibilities were grossly exaggerated.
*shrug* whatever.

FYI: I maintain RPM5, and have several (like almost 10 ;-) years of experience dealing with Poky/Yocto packaging issues.

Meanwhile I suspect that capsule is only the first, not the last, RPM5 package that you will see. Ripping out the "exact_size" test will solve the issue until there is more info or a better fix available.

**Comment by anselmolsm** [comment]
@n3npq, thanks for your comments.

It's just to add more info:
A possible conflict between package generated by rpm5 x rpm4 was one of our first hypothesis, however the same issue happens with signed packages generated by rpm4 (for the yocto-2.3 based system):

```
bash-4.3# rpm5 -qp --yaml intel-aero-repo-1.4-r0.corei7_64.rpm | grep Rpmversion
  Rpmversion: 4.13.90
```


rpm 4 with the commit linked by @lucasdemarchi above fails with the the following error:
```
bash-4.3# rpm -qlp intel-aero-repo-1.4-r0.corei7_64.rpm 
error: intel-aero-repo-1.4-r0.corei7_64.rpm: signature region 62: tag number mismatch 8 ril 6 dl 4772 rdl 16

error: intel-aero-repo-1.4-r0.corei7_64.rpm: not an rpm package (or package manifest)
```

Both [intel-aero-repo](https://github.com/rpm-software-management/rpm/files/1166904/intel-aero-repo-1.4-r0.corei7_64.rpm.zip) package and the [public key](https://github.com/rpm-software-management/rpm/files/1166908/public-gpg-key.txt) are attached for reference.

**Comment by n3npq** [comment]
btw, this command is sufficient to identify the rpm version that built the package:
`rpm -qp --qf '%{rpm version}\n' intel-aero-repo-1.4-r0.corei7_64.rpm`

It looks the  arguments passed used in the exact_size check are incorrect (the "ril" and "rdl" values in the error message seem wrong).

Commenting out this code permits reading the package:
```
--- a/lib/header.c
+++ b/lib/header.c
@@ -1840,7 +1840,9 @@ static rpmRC hdrblobVerifyRegion(rpmTagVal regionTag, int 
        rasprintf(buf,
                _("region %d: tag number mismatch %d ril %d dl %d rdl %d\n"),
                regionTag, blob->il, blob->ril, blob->dl, blob->rdl);
+#ifdef HACK
        goto exit;
+#endif
     }
 
     blob->regionTag = regionTag;
```

The real fix is somewhere else, in the value passed to the exact_size boolean argument, but will have the same effect of not checking the index/data length of a signature header immutable region.

**Comment by n3npq** [comment]
Other nitpicks  in the geeky error message: the check has nothing to do with a "tag number mismatch", and lacks an "il" identifier that might have helped diagnose the problem slightly more rapidly.

OTOH, I'm not sure anyone can diagnose a problem in the binary header format without a reproducer and studying the code.

**Comment by pmatilai** [comment]
The check is not changing what immutable region is, it's there because traditionally headers coming off packages on the disk have not had anything outside the immutable region (and yes the message doesn't make a very good job of expressing that, shrug) - with the exception of:
- repackaged packages, which we don't support (and this makes it explicit)
- fuzzed packages, for which this shuts the door earlier

And now this, which I take it is a relatively recent change in rpm5.

Rpm5 compatibility - or incompatibility - are not goals for rpm.org. I suggest you build the packages with the rpm version you plan to install them with.

FWIW, the missing "il" added to the message.

**Comment by n3npq** [comment]
Note that a package produced by rpm 4.13 reproduced the problem.

**Comment by pmatilai** [comment]
Oh, sorry for missed that. But now the question is, what rpm version is used to *sign* that package? 

It doesn't seem like any rpm.org upstream version because deleting the signature (with rpm <= 4.13.0) corrupts the package beyond recognition of those older versions too:
```
[pmatilai@sopuli rpm-4.13.x]$ cp ~/Downloads/intel-aero-repo-1.4-r0.corei7_64.rpm /tmp/
[pmatilai@sopuli rpm-4.13.x]$ ./rpmkeys -Kv /tmp/intel-aero-repo-1.4-r0.corei7_64.rpm 
/tmp/intel-aero-repo-1.4-r0.corei7_64.rpm:
    Header V3 RSA/SHA1 Signature, key ID bb0396db: NOKEY
    Header SHA1 digest: OK (3066ac48fef77e939547216732d92bb047eb85b6)
    V3 RSA/SHA1 Signature, key ID bb0396db: NOKEY
    MD5 digest: OK (cf96418cf798c2ef1351fb012e7a6b7e)
[pmatilai@sopuli rpm-4.13.x]$ ./rpmsign --delsign /tmp/intel-aero-repo-1.4-r0.corei7_64.rpm
/tmp/intel-aero-repo-1.4-r0.corei7_64.rpm:
[pmatilai@sopuli rpm-4.13.x]$ ./rpmkeys -Kv /tmp/intel-aero-repo-1.4-r0.corei7_64.rpm 
error: /tmp/intel-aero-repo-1.4-r0.corei7_64.rpm: sigh data: BAD, no. of  bytes(8900) out of range
```

If this is signed with an rpm.org release, a reproducer please.

**Comment by pmatilai** [comment]
FWIW that --delsign corruption is of course a bug in <= 4.13.0 someplace, if it can't handle it should at least refuse to touch it. That corruption longer happens in 4.13.90 (once patched to accept the package in the first place) so it's gotten fixed somewhere along the road, whatever it is. Bisecting that is a bit of a PITA as there are other related bugs present during development...

**Comment by anselmolsm** [comment]
An update, I've contacted the maintainers of the internal signing service to obtain information about the rpm version used by their system to sign packages. I'll post an update as soon as I hear from them.

**Comment by pmatilai** [comment]
There's now another similar bug at https://bugzilla.redhat.com/show_bug.cgi?id=1514190, Intel signed packages in that case as well, with the same problem: the actual signature is placed outside signature header immutable region. I'm not aware of any rpm.org version doing that. @anselmolsm - any news on that front?

@n3npq , as for the original report on the rpm5 built capsule package: the "offending" tag is RPMSIGTAG_PADDING and I was about to ask why it is outside the immutable region, but looking at the code it doesn't seem to be intentional:

```
    /* Reallocate the signature header into one contiguous region. */
    sigh = headerReload(sigh, RPMTAG_HEADERSIGNATURES);

   [...]

        he->tag = (rpmTag) RPMSIGTAG_PADDING;
        he->t = RPM_BIN_TYPE;
        he->p.ui8p = b;
        he->c = nb;
        xx = headerPut(sigh, he, HEADERGET_SIGHEADER);
        sigh = headerReload(sigh, RPMTAG_HEADERSIGNATURES);
```
So it seems the second headerReload() fails to pull the padding tag into the region, for whatever reason.

**Comment by pmatilai** [comment]
...to continue on the padding tag placement, this is what I see in the capsule package signature header:
```
Index entries: 7 (112 bytes)
Data size: 792
Header size: 904
Tag #0
        tagno:    62 (Headersignatures)
        type:      7 (blob)
        offset:  152
        count:    16
[...]
Tag #5
        tagno:  1007 (Payloadsize)
        type:      4 (int32)
        offset:  148
        count:     1
Tag #6
        tagno:  1073741823 ((unknown))
        type:      7 (blob)
        offset:  168
        count:   624
```
So the region trailer is between RPMSIGTAG_PAYLOADSIZE and RPMSIGTAG_PADDING data, not last as I'd expect from the code.

**Comment by n3npq** [comment]
The signature padding tag cannot be part of an immutable region because its size/value can/will change. There is also little reason to include padding within plaintext used for digests/signatures

As for ordering: you chose a different value for the padding tag than what is implemented in rpm5. So a different order is expected when sorting. The value used in rpm5 was chosen to always sort after all other defined tags that might appear in signature headers in order to function as "free space" padding.

**Comment by n3npq** [comment]
@pmatlai: you are claiming "might well be a bug" in rpm5 at bz#1514190 without supplying details.

Either supply sufficient details (or a patch preferred) or stop claiming FUD, entirely your call.

**Comment by pmatilai** [comment]
Ehm, maybe we have a terminology problem here. When I talk about signature header immutable region, I mean HEADER_SIGNATURES region. That's not used for digests or signatures and when adding/removing signatures the signature header changes more fundamentally than just padding adjustments. So yes if you look at it in terms of what might change, only the build-time added digests and sizes etc should go into the region and actual signatures etc that might be added over time should go outside it, but rpm has "forever" reloaded the signature into a region after each signature add/remove and all. Padding only changes when signatures are added or removed, so technically it can just as well be in the region (or not). So yes the region doesn't mean a whole lot in the signature header...

As for the bug allegation, the details are right there in my previous posts: the code (copy-pasted from rpm5 as you must've noticed) calls headerReload() after adding the padding, which I read as "make it a single (immutable) region". And despite that, the padding ends up being outside the region. I'm not arguing whether it should be in the region or not, just observing that the result doesn't match what was indicated by the code, which to me seems like a possible bug someplace, which I'm pointing out as in "you might want to look at that", no ill-will / FUD / whatever intended. If I'm mistaken then by golly apologies for speculating the possibility of lurking bugs in a piece of software.

**Comment by pmatilai** [comment]
Oh okay, took a dog-walk for it to register: in rpm5 RPMSIGTAG_PADDING is used for the actual padding between the signature and main header, which is different from rpm.org's RPMSIGTAG_RESERVEDSPACE. From the initial description of #183 I assumed it was essentially the same and, I admit, didn't bother with the other details you provided later. So it's different, in a number of ways, my bad for assuming otherwise.

But in that case, calling headerReload() again after adding the padding tag seems wrong, it's just that the second call doesn't actually seem to do much anything so it ends up being harmless. This is true of rpm.org's headerReload() as well BTW - it only works once to actually pull it all into that single contiguous region. I seem to recall running into that once or twice but never got around to look into it just then, there hasn't been much of a reason to bother.

**Comment by pmatilai** [comment]
Anyway... the check disabled for the signature header in commit 34c2ba3c6a80a778cdf2e42a9193b3264e08e1b3, thanks for the patience everybody.
There will be an update to 4.14.x sooner than later, this will be included.

**Comment by n3npq** [comment]
Several points:

1) Your check for repackaged packages is what is running afoul of (what is obscurely known as) a "dribble", a tag appended to an immutable region, which is commonly found in repackaged packages (with a doubly linked upgrade chain tag appended) and in rpmdb headers (signature tags, and local installation data). You are correct that no *.rpm file has ever had a "dribble" appended, but there is nothing in a Header that prevents a "dribble" from being used.

2) headerReload() @rpm5.org does more than copy tags into a contiguous malloc. In fact, an anonymous mmap(2) is used, and mprotect(2) PROT_READ is done to prevent *all* changes to *any* header data returned as a pointer to the user. This also makes the pages used to store a Header RO, eliminates the silliness of HEADER_MINMEM permitting access directly into header data, and (as side effect) stops the overhead of swabbing all integer data on every headerLoad. Instead, data is swabbed during headerGet() returned as a malloc'd copy, to itemize just a few of the differences that prevent reasoning about the behavior from an identically named "headerReload" symbol.

3) RPMSIGTAG_PADDING can be used not only to reserve space for signatures, but also to insert block alignment between signature <-> metadata headers (useful for byte ranged HTTP), and (if also appended to the metadata header( between metadata <-> payload. Of course having the ability to do HTTP byte range retrievals suffers from "legacy compatibility" in a sense (backward compatibility is ensured by the format, but the "feature" cannot be assumed to exist in any given *.rpm without some other test).

There are yet other reasons to use a "dribble" in a *.rpm package, including unifying (by nesting immutable regions) signature <-> metadata Header's into a single blob.

Details are left in my imagination ;-)

**Comment by n3npq** [comment]
TL;DR The 2nd headerReload() appends RPMSIGTAG_PADDING into a contiguous PROT_READ block.
